Type housekeeping.

paulp opened the pull request:
Moved ListOfNil somewhere more generally accessible. No
reason the compiler should hoard it for itself. Flitted
to a few locations with constructs like ".head.head" and
".tail.head" looking for code which could be rewritten. Found
some, admittedly not always making use of ListOfNil.

Made overdue moves of ConstantType(Constant(true|false|null))
to vals in Definitions.

paulp commented:
Review by @retronym

som-snytt commented:
Am I the only one thinking it sounds like the basis for a skit with "The Knights Who Say Nil"?

retronym commented on src/reflect/scala/reflect/internal/Definitions.scala:
```diff
@@ -544,7 +548,7 @@ trait Definitions extends api.StandardDefinitions {
          def MacroContextPrefix           = MacroContextClass.map(sym => getMemberMethod(sym, nme.prefix))
          def MacroContextPrefixType       = MacroContextClass.map(sym => getTypeMember(sym, tpnme.PrefixType))
-         def MacroContextUniverse         = MacroContextClass.map(sym => getMemberMethod(sym, nme.universe))
+         lazy val MacroContextUniverse    = MacroContextClass.map(sym => getMemberMethod(sym, nme.universe))
```
Members should not be `vals` as they can acquire new symbols in different runs if, at least, the enclosing symbol switches from one based on class files to one based on source files. This happens in the IDE. See https://issues.scala-lang.org/browse/SI-7678

paulp replied:
I was afraid there was some reason like that. No pattern matching, I may as well be writing in java.

retronym replied:
You can keep the pretty match at the cost of a local val.

paulp replied:
Already paid the piper.

retronym commented:
LGTM, thanks!

paulp committed:
Type housekeeping.

Moved ListOfNil somewhere more generally accessible. No
reason the compiler should hoard it for itself. Flitted
to a few locations with constructs like ".head.head" and
".tail.head" looking for code which could be rewritten. Found
some, admittedly not always making use of ListOfNil.

Made overdue moves of ConstantType(Constant(true|false|null))
to vals in Definitions.

retronym commented:
LGTM, still, after the squash.

paulp commented:
Great, I'll take it.
